### PR TITLE
Use Resources in Menu and rename Menu -> List

### LINKED ISSUE #1162: Convert Menu to use resources and rename to List
Now that resources / middleware have been released in framework we can convert the `Menu` widget to use these. 
Furthermore, we should rename `Menu` to `List` as it is more fitting.
This issue follows on from discussion here: https://github.com/dojo/widgets/issues/1010

### PR DESCRIPTION
**Pull request by tomdye**
**Type:** bug / feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/framework/blob/master/STYLE.md)
* [x] Unit tests are included in the PR
* [x] For new widgets, an entry has been added to the `.dojorc`
* [x] Any widget variant uses `theme.compose` like [this](https://github.com/dojo/widgets/issues/847)
* [x] WidgetProperties are exported

**Description:**

- Renames `menu` to `list`
- uses resources within list
- updates `select`, `context-menu` and `timepicker`
- updates examples

Resolves #1162 

Requires https://github.com/dojo/framework/pull/701 and https://github.com/dojo/framework/pull/703 and framework release.

### CODE REVIEW

File: src/examples/src/widgets/menu/FetchedResource.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
+import { create, tsx } from '@dojo/framework/core/vdom';
+import List from '@dojo/widgets/list';
+import icache from '@dojo/framework/core/middleware/icache';
+import { createResource, createTransformer, DataTemplate } from '@dojo/widgets/common/resource';
+
+const fetcher = async (options: any) => {
```
Comment (agubler) [inline src/examples/src/widgets/menu/FetchedResource.tsx]: Might as well make this non remote for the examples (as my server is slow and if it breaks the examples break)

File: src/examples/src/widgets/menu/MemoryResource.tsx [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+const animals = [
+	{ name: 'cat', label: 'cat: feline' },
+	{ name: 'dog', label: 'dog: kanine', disabled: true },
+	{ name: 'mouse', label: 'mouse: rodent' },
+	{ name: 'rat', label: 'rat: rodent' }
+];
+
+export function createMemoryTemplate<S = void>(): DataTemplate<S> {
+	return {
+		read: ({ query }, put, get) => {
```
Comment (agubler) [inline src/examples/src/widgets/menu/MemoryResource.tsx]: Is this the part you want to provide a helper for?

File: src/theme/dojo/index.ts [hunk 3]
```diff
@@ -75,10 +75,10 @@ export default {
 	'@dojo/widgets/icon': icon,
 	'@dojo/widgets/label': label,
 	'@dojo/widgets/listbox': listbox,
-	'@dojo/widgets/list-box-item': listboxItem,
+	'@dojo/widgets/list-box-item': listItem,
 	'@dojo/widgets/loading-indicator': loadingIndicator,
 	'@dojo/widgets/menu-item': menuItem,
-	'@dojo/widgets/menu': menu,
+	'@dojo/widgets/menu': list,
```
Comment (agubler) [inline src/theme/dojo/index.ts]: this should be `widgets/list`

File: src/theme/material/index.ts [hunk 4]
```diff
@@ -74,9 +74,9 @@ export default {
 	'@dojo/widgets/icon': icon,
 	'@dojo/widgets/label': label,
 	'@dojo/widgets/listbox': listbox,
-	'@dojo/widgets/list-box-item': listBoxItem,
+	'@dojo/widgets/list-box-item': listItem,
 	'@dojo/widgets/loading-indicator': loadingIndicator,
-	'@dojo/widgets/menu': menu,
+	'@dojo/widgets/menu': list,
```
Comment (agubler) [inline src/theme/material/index.ts]: same comment as above

File: src/examples/src/widgets/menu/MemoryResource.tsx (hunk 2 above)
Comment (tomdye) [reply src/examples/src/widgets/menu/MemoryResource.tsx]: yes, will need to support filtering too

File: src/theme/dojo/index.ts (hunk 3 above)
Comment (tomdye) [reply src/theme/dojo/index.ts]: thanks

File: src/theme/material/index.ts (hunk 4 above)
Comment (tomdye) [reply src/theme/material/index.ts]: thanks

File: src/examples/src/widgets/menu/FetchedResource.tsx (hunk 1 above)
Comment (tomdye) [reply src/examples/src/widgets/menu/FetchedResource.tsx]: no worries, can do

File: src/examples/src/widgets/list/FetchedResource.tsx [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		<virtual>
+			<List
+				resource={resource}
+				transform={transformer}
+				onValue={(value: string) => {
+					icache.set('value', value);
+				}}
+				itemsInView={10}
+			/>
+			<p>{`Clicked on: ${icache.getOrSet('value', '')}`}</p>{' '}
```
Comment (rorticus) [inline src/examples/src/widgets/list/FetchedResource.tsx]:
```suggestion
			<p>{`Clicked on: ${icache.getOrSet('value', '')}`}</p>
```

File: src/time-picker/index.tsx [hunk 6]
```diff
@@ -12,6 +12,19 @@ import { createICacheMiddleware } from '@dojo/framework/core/middleware/icache';
 import { Keys } from '../common/util';
 import bundle from './nls/TimePicker';
 import i18n from '@dojo/framework/core/middleware/i18n';
+import { createResource, DataTemplate } from '@dojo/framework/core/resource';
+
+function createMemoryTemplate<S = void>(): DataTemplate<S> {
+	return {
+		read: ({ query }, put, get) => {
+			let data: any[] = get();
+			put(0, data);
```
Comment (maier49) [inline src/time-picker/index.tsx]: Why do we need to `get` and then immediately `put` the same data?

File: src/examples/src/widgets/context-menu/Basic.tsx [hunk 7]
```diff
@@ -11,11 +13,16 @@ export default factory(function Basic({ middleware: { icache } }) {
 		'This is some text that has a context menu. Right click to view the context menu and take an action on selected text'
 	);
 	const printedText = icache.getOrSet('printedText', '');
+	const memoryTemplate = createMemoryTemplate();
 
 	return (
 		<virtual>
 			<ContextMenu
-				options={options}
+				resource={{
```
Comment (maier49) [inline src/examples/src/widgets/context-menu/Basic.tsx]: I'm a bit out of loop on the resource stuff, but it feels like a bit complex to have to create a template to create a resource just to pass an array of data.

File: src/list/README.md [hunk 8]
```diff
@@ -1,18 +1,18 @@
-# @dojo/widgets/menu
+# @dojo/widgets/list
```
Comment (rorticus) [inline src/list/README.md]: Should this talk about, or link to, resources?

File: src/time-picker/index.tsx [hunk 9]
```diff
@@ -12,6 +12,19 @@ import { createICacheMiddleware } from '@dojo/framework/core/middleware/icache';
 import { Keys } from '../common/util';
 import bundle from './nls/TimePicker';
 import i18n from '@dojo/framework/core/middleware/i18n';
+import { createResource, DataTemplate } from '@dojo/framework/core/resource';
+
+function createMemoryTemplate<S = void>(): DataTemplate<S> {
```
Comment (rorticus) [inline src/time-picker/index.tsx]: Can this not be imported like the other createMemoryTemplate?

File: src/examples/src/widgets/context-menu/Basic.tsx (hunk 7 above)
Comment (tomdye) [reply src/examples/src/widgets/context-menu/Basic.tsx]: yes, agreed. We will be working on providing suitable helpers for these to make it a lot easier and simpler.

File: src/time-picker/index.tsx (hunk 6 above)
Comment (tomdye) [reply src/time-picker/index.tsx]: The `read` function gets passed  both a getter and a setter function. In the case of a memory template, the first request may only call `read` for the first few results, but we wish to get all of the data and set it back again against the query that was passed. This simple template is not the best example of this functionality really but it _does_ begin to make more sense when filters etc are concerned. Basically, for a select etc, it ensure that all the items are available and have been sideloaded into the widget meaning items can be navigated to via `a-z` press etc.

File: src/time-picker/index.tsx (hunk 9 above)
Comment (tomdye) [reply src/time-picker/index.tsx]: this is an actual widget, the helper is in the examples dir until we create a better one in framework

File: src/list/README.md (hunk 8 above)
Comment (tomdye) [reply src/list/README.md]: i can certainly raise a follow up to discuss etc

File: src/examples/src/widgets/context-menu/Basic.tsx (hunk 7 above)
Comment (tomdye) [reply src/examples/src/widgets/context-menu/Basic.tsx]: have made these simpler

**Pull request merged**